build: x86 ASan build broken "error: inline assembly requires more registers than available"

Dockerfile to reproduce: https://gist.github.com/dergoegge/fc97743028f60719759b5498f5f022bf

Building with `clang-19` and `-fsantize=address` fails with:

```
7.574 /bitcoin/src/secp256k1/src/scalar_4x64_impl.h:356:5: error: inline assembly requires more registers than available
7.574   356 |     "movq 32(%%rsi), %%r11\n"
7.574       |     ^
7.575 /bitcoin/src/secp256k1/src/scalar_4x64_impl.h:356:5: error: inline assembly requires more registers than available
7.575 /bitcoin/src/secp256k1/src/scalar_4x64_impl.h:356:5: error: inline assembly requires more registers than available
7.643 [ 24%] Building CXX object src/CMakeFiles/bitcoin_consensus.dir/uint256.cpp.o
7.722 3 errors generated.
7.738 gmake[3]: *** [src/secp256k1/src/CMakeFiles/secp256k1.dir/build.make:76: src/secp256k1/src/CMakeFiles/secp256k1.dir/secp256k1.c.o] Error 1
7.738 gmake[2]: *** [CMakeFiles/Makefile2:1020: src/secp256k1/src/CMakeFiles/secp256k1.dir/all] Error 2
```

Since this looks secp256k1 related I tried reproducing on a separate secp build but couldn't. This could very well be a compiler/linker bug.

I am surprised that I'm the only one running into this, we don't have a x86 ASan job anywhere? I think I first observed this sometime mid last year.

>  Can you try if -fomit-frame-pointer fixes it?

Still the same issue.

```
root@7069ec4b9a4e:/# clang-19  -DUSE_ASM_X86_64=1 -std=c90   -o secp256k1.c.o -c /bitcoin/src/secp256k1/src/secp256k1.c -fsanitize=address -fomit-frame-pointer
In file included from /bitcoin/src/secp256k1/src/secp256k1.c:28:
In file included from /bitcoin/src/secp256k1/src/scalar_impl.h:20:
/bitcoin/src/secp256k1/src/scalar_4x64_impl.h:356:5: error: inline assembly requires more registers than available
  356 |     "movq 32(%%rsi), %%r11\n"
      |     ^
/bitcoin/src/secp256k1/src/scalar_4x64_impl.h:356:5: error: inline assembly requires more registers than available
/bitcoin/src/secp256k1/src/scalar_4x64_impl.h:356:5: error: inline assembly requires more registers than available
3 errors generated.
```

Not using `-DUSE_ASM_X86_64=1` obviously works. Could be a work around once https://github.com/bitcoin/bitcoin/pull/31379 is merged.

--- a/src/crypto/sha256_sse4.cpp
+++ b/src/crypto/sha256_sse4.cpp
@@ -13,12 +13,16 @@
 namespace sha256_sse4
 {
 void Transform(uint32_t* s, const unsigned char* chunk, size_t blocks)
-#if defined(__clang__) && !defined(__OPTIMIZE__)
+#if defined(__clang__)
   /*
   clang is unable to compile this with -O0 and -fsanitize=address.
-  See upstream bug: https://github.com/llvm/llvm-project/issues/92182
+  See upstream bug: https://github.com/llvm/llvm-project/issues/92182.
+  This also fails to compile with -O2, -fcf-protection & -fsanitize=address.
+  See https://github.com/bitcoin/bitcoin/issues/31913.
   */
+#if __has_feature(address_sanitizer)
   __attribute__((no_sanitize("address")))
+#endif
 #endif
 {
     static const uint32_t K256 alignas(16) [] = {

